app/src/main/java/com/example/currencyconverter/ui/di/ServiceLocator.kt

```diff
@@ -0,0 +1,41 @@
+package com.example.currencyconverter.ui.di
+
+import android.content.Context
+import com.example.currencyconverter.data.database.AppDatabase
+import com.example.currencyconverter.data.network.NetworkService
+import com.example.currencyconverter.data.repository.CurrencyRepository
+import com.example.currencyconverter.ui.converter.utils.ValueInputValidator
+
+object ServiceLocator {
+
+    private var database: AppDatabase? = null
+
+    private val networkService by lazy {
+        NetworkService()
+    }
+
+    @Volatile
+    var currencyRepository: CurrencyRepository? = null
+
+    fun provideCurrencyRepository(context: Context): CurrencyRepository {
+        synchronized(this) {
+            return currencyRepository ?: createCurrencyRepository(context)
+        }
+    }
+
+    private fun createCurrencyRepository(context: Context): CurrencyRepository {
+        val database = database ?: createDatabase(context)
+        return CurrencyRepository(database.databaseDao, networkService.getCurrencyApi())
+    }
+
+    private fun createDatabase(context: Context): AppDatabase {
+        val result = AppDatabase.getInstance(context)
+        database = result
+        return result
+    }
+
+    fun provideValueInputValidator(): ValueInputValidator {
+        return ValueInputValidator()
```

Comment:
It will be created from scratch on every 'provideValueInputValidator()' call. Is this what you wanted?